Add performance testing

sdonkov opened the pull request:
Adding performance testing using jol.

sdonkov committed:
add jol dependency, add dictionary, compare prefixTree vs hashSet vs treeSet

dmitankin reviewed (commented):
It makes sense to reduce the log level for the tests to `debug`; otherwise, the log is flooded.

What I mean by performance tests is [microbenchmarking](https://www.baeldung.com/java-microbenchmark-harness). The goal is to compare the available functionality between data-structures in terms of time complexity too, not only memory one.

sdonkov committed:
add jmh dependencies, add benchmark tests, change root level to debug

dmitankin commented on src/test/java/com/github/sdonkov/PrefixTreeTest.java:
```diff
@@ -1,28 +1,76 @@
+        words.forEach(word -> prefixTree.add(word));
+    }
+
+    @Test
```
We definitely do not want to run this along with the unit tests.
It should be a separate command that triggers perf testing so that development is not hindered. While changing the code, the dev often needs a quick check whether something is broken. The perf tests are something to run sparingly, e.g. prior to posting a pull request that has the potential to regress or improve on the performance front.

It does make sense to move the perf testing to a separate file. Files are useful to avoid collisions during development. Ideally, you should have a single reason to change a file. This file contains tests for the API contract of the PrefixTree. Changing the contract should be the only reason to touch it.

dmitankin commented on src/test/java/com/github/sdonkov/PrefixTreeTest.java:
```diff
@@ -1,28 +1,76 @@
+        LOGGER.trace(GraphLayout.parseInstance(prefixTree).totalSize() + " prefixTree");
+    }
+
+    @Benchmark
+    public void benchmarkHashSet() throws URISyntaxException, IOException {
+        words = Files.readAllLines(Paths.get(PrefixTreeTest.class.getResource("/dictionary.txt").toURI()));
+        Set<String> hashSet = new HashSet<>(words);
+        LOGGER.trace(GraphLayout.parseInstance(hashSet).totalSize() + " hashSet");
+
+    }
+
+
+    @Benchmark
+    public void benchmarkTreeSet() throws URISyntaxException, IOException {
+        words = Files.readAllLines(Paths.get(PrefixTreeTest.class.getResource("/dictionary.txt").toURI()));
+        Set<String> treeSet = new TreeSet<>(words);
+        LOGGER.trace(GraphLayout.parseInstance(treeSet).totalSize() + " treeSet");
```
Test code is like any other code and benefits from [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) too.

dmitankin commented on src/test/java/com/github/sdonkov/PrefixTreeTest.java:
```diff
@@ -1,28 +1,76 @@
+        prefixTree = new PrefixTree();
+        words = Files.readAllLines(Paths.get(PrefixTreeTest.class.getResource("/dictionary.txt").toURI()));
         words.forEach(word -> prefixTree.add(word));
+        LOGGER.trace(GraphLayout.parseInstance(prefixTree).totalSize() + " prefixTree");
```
This is not what we want from the microbenchmark.
We need to know how fast the structure works. Currently we have two methods exposed as functionality - `add` and `contains`. I say we test only `contains`. The test should be something like:
0. Setup the test subject by adding all words from the dictionary. Pick randomly 10 words from the dictionary. This step should be performed prior to running the benchmark.
1. Test how fast does the structure checks whether all the words are contained and repeat the test many times to get an average. The second part is provided by the microbenchmark framework.

Outside of the benchmark, it'd be very useful to print how much memory each tested structure consumes. There's no need to check many times.

sdonkov committed:
add PrefixTreeBenchmark, remove unused imports in PrefixTreeTest

sdonkov committed:
add blackhole in benchmark methods, change plugin in pom

dmitankin commented on src/test/java/com/github/sdonkov/PrefixTreeBenchmark.java:
```diff
@@ -0,0 +1,57 @@
+import java.net.URISyntaxException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.*;
```
Is this decided by your IDE? I guess there's a threshold of imported classes over which it decides to import everything.
It'd be best to turn it off and import everything explicitly. Wildcard imports are fine if they are for test stuff like the annotations here or assert methods in unit tests, but hurt readability otherwise.

dmitankin commented on pom.xml:
```diff
@@ -44,10 +44,52 @@
+        <dependency>
+            <groupId>org.openjdk.jmh</groupId>
+            <artifactId>jmh-core</artifactId>
+            <version>1.36</version>
```
Ideally this version would be a property to make upgrading the dependency a single line of change; otherwise, there's a risk that someone will update this dependency but leave the one below to an older version.

dmitankin commented on src/test/java/com/github/sdonkov/PrefixTreeBenchmark.java:
```diff
@@ -0,0 +1,57 @@
+        treeSet = new TreeSet<>(words);
+        hashSet = new HashSet<>(words);
+        LOGGER.debug(GraphLayout.parseInstance(prefixTree).totalSize() + " tree");
+//        LOGGER.debug(GraphLayout.parseInstance(hashSet).totalSize() + " hashSet");
+//        5377400 hashSet
+//        LOGGER.debug(GraphLayout.parseInstance(treeSet).totalSize() + " treeSet");
+//        5317968 treeSet
```
Ah, sorry, I forgot we altready have these structures present and populated. I guess it doesn't make sense then to have these commented out since changing the contents of the dictionary will render the numbers obsolete.

dmitankin commented on src/test/java/com/github/sdonkov/PrefixTreeTest.java:
```diff
@@ -18,10 +18,11 @@ public class PrefixTreeTest {
     private static List<String> words;
     private static PrefixTree prefixTree;
 
+
```
No need for this new line.

dmitankin commented on src/test/java/com/github/sdonkov/PrefixTreeBenchmark.java:
```diff
@@ -0,0 +1,57 @@
+        Collections.shuffle(words);
+        treeSet = new TreeSet<>(words);
+        hashSet = new HashSet<>(words);
+        LOGGER.debug(GraphLayout.parseInstance(prefixTree).totalSize() + " tree");
```
shoudln't this print `PrefixTree` instead of `tree`?
Actually, it'd be best if there's a method in this class that prints the info for an object. We can then call it three times to ensure consistent output.
We'd be able to use it for other implentations down the road.

dmitankin commented on src/test/java/com/github/sdonkov/PrefixTreeBenchmark.java:
```diff
@@ -0,0 +1,57 @@
+@Fork(value = 1)
+@Warmup(iterations = 3)
+@State(Scope.Benchmark)
+public class PrefixTreeBenchmark {
```
This class is not particular to PrefixTree and we'll use it for other implementations too. Simply `Benchmark` will do, I guess.

dmitankin commented on src/test/java/com/github/sdonkov/PrefixTreeBenchmark.java:
```diff
@@ -0,0 +1,57 @@
+    private static PrefixTree prefixTree;
+    private static Set<String> treeSet;
+    private static Set<String> hashSet;
+    private final static Logger LOGGER = LogManager.getLogger(PrefixTree.class);
```
Note that this will print log statements with the "PrefixTree" logger. In this case it'd be best to use the current class' logger.

dmitankin commented on src/test/java/com/github/sdonkov/PrefixTreeBenchmark.java:
```diff
@@ -0,0 +1,57 @@
+        prefixTree = new PrefixTree();
+        words = Files.readAllLines(Paths.get(PrefixTreeTest.class.getResource("/dictionary.txt").toURI()));
+        words.forEach(word -> prefixTree.add(word));
+        Collections.shuffle(words);
```
It looks like shuffling is performed on purpose before setting up the other two structures. It'd be best to move it at the end.

sdonkov committed:
disable wildcard imports,

add jmh version property in pom,

add sizeinfo method,

rename benchmark class,

change logger,

move shuffle

dmitankin commented on src/test/java/com/github/sdonkov/Benchmark.java:
```diff
@@ -0,0 +1,70 @@
+        words.forEach(word -> bh.consume(prefixTree.contains(word)));
+    }
+
+    //    @Benchmark
```
Apply the FQN here too or simply rename this class to something else that won't clash with the annotation name.

dmitankin commented on src/test/java/com/github/sdonkov/Benchmark.java:
```diff
@@ -0,0 +1,70 @@
+        words.forEach(word -> prefixTree.add(word));
+        treeSet = new TreeSet<>(words);
+        hashSet = new HashSet<>(words);
+        sizeInfo(prefixTree);
```
great, let's use it for the two sets too

sdonkov committed:
rename benchmark class, add sizeinfo for the two sets

dmitankin commented on src/test/java/com/github/sdonkov/DataStructureBenchmark.java:
```diff
@@ -0,0 +1,72 @@
+        LOGGER.debug(GraphLayout.parseInstance(object).totalSize() + " " + object.getClass().getSimpleName());
+    }
+
+    @org.openjdk.jmh.annotations.Benchmark
```
You can now import the annotation class and use its simple name

sdonkov committed:
update imports

dmitankin reviewed (approved):
It'd be nice to post the results of a perf run from your machine in the description of the merge commit.

The numbers I got for memory consumption are:
```
12649016 PrefixTree
5788608 TreeSet
5848040 HashSet
```

And the results from the benchmark:
```
Benchmark                                   Mode  Cnt   Score   Error  Units
DataStructureBenchmark.benchmarkPrefixTree  avgt    5  33,165 ▒ 1,300  ms/op
```
